tendermint/src/lite_impl/signed_header.rs

```diff
@@ -58,69 +59,107 @@ impl lite::Commit for block::signed_header::SignedHeader {
-                    votes.push(vote);
+        if let Some(val_addr) = commit_sig.validator_address {
+            if let Some(sig) = commit_sig.signature.clone() {
+                let vote = vote::Vote {
+                    vote_type: vote::Type::Precommit,
+                    height: commit.height,
+                    round: commit.round,
+                    block_id: Option::from(commit.block_id.clone()),
+                    timestamp: commit_sig.timestamp,
+                    validator_address: val_addr,
+                    validator_index: i as u64,
+                    signature: sig,
+                };
+                votes.push(vote);
+            }
+        }
+    }
+    votes
+}
+
+// TODO: consider moving this into commit_sig.rs instead and making it pub
+impl block::commit_sig::CommitSig {
+    fn validate(&self, vals: &Set) -> Result<(), Error> {
+        match self.block_id_flag {
+            BlockIDFlag::BlockIDFlagAbsent => {
+                if self.validator_address.is_some() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "validator address is present for absent CommitSig {:#?}",
+                        self
+                    );
                 }
+                if self.signature.is_some() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "signature is present for absent CommitSig {:#?}",
+                        self
+                    );
+                }
+                // TODO: deal with Time
+                // see https://github.com/informalsystems/tendermint-rs/pull/196#discussion_r401027989
             }
-            None => {
-                fail!(
-                    Kind::ImplementationSpecific,
-                    "validator address missing in commit_sig {:#?}",
-                    commit_sig
-                );
+            BlockIDFlag::BlockIDFlagCommit | BlockIDFlag::BlockIDFlagNil => {
+                if self.validator_address.is_none() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "missing validator address for non-absent CommitSig {:#?}",
+                        self
+                    );
+                }
+                if self.signature.is_none() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "missing signature for non-absent CommitSig {:#?}",
+                        self
+                    );
+                }
+                // TODO: this last check is only necessary if we do full verification (2/3) but the
+                // above checks should actually happen always (even if we skip forward)
+                //
+                // returns ImplementationSpecific error if it detects a signer
+                // that is not present in the validator set:
+                if let Some(val_addr) = self.validator_address {
```

Comment:
Maybe move this check into a separate function which is only called in case of full verification?

Reply:
Will do in a followup PR.